### PR TITLE
Add upgrading and version management documentation

### PR DESCRIPTION
**Pull request by tedsuo**
Long ago, we defined upgrading procedures for OpenTelemetry, in [OTEP 147](https://github.com/open-telemetry/oteps/blob/main/text/0147-upgrade-procedures.md). However, we never added this language to the specification.

This pull request adds the OTEP to the specification, verbatim, except for an addition at the top which clearly spells out dependency management for application developers, library maintainers, and SDK implementors.

It has been two years, so I am open to changes to this document. Both for clarity, and to conform to how we currently practice upgrading and dependency management.

### CODE REVIEW

File: specification/upgrading.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+API implementations are expected to always target the latest version of the API. When a new version of the API is released, a version of the SDK which supports the API is released in tandem. New versions of the API are not expected to support older versions of the SDK.
+
+### SDK changes
+
+Bugs fixes, security patches, and performance improvements are released as patch versions of the SDK. Support for new versions of the API are released as minor versions. New Plugin Interfaces and configuration options are also released as minor versions.
+
+Breaking changes to Plugin Interfaces are handled through deprecation. Instead of breaking a Plugin Interface, a new interface is created and the existing interface is marked as deprecated. Plugins which target the deprecated interface continue to work, and the SDK provides default implementations of any missing functionality. After one year, the deprecated Plugin Interface is removed as a major version release of the SDK.
```
Comment (trask) [inline specification/upgrading.md:62]:
```suggestion
Breaking changes to Plugin Interfaces are handled through deprecation. Instead of breaking a Plugin Interface, a new interface is created and the existing interface is marked as deprecated. Plugins which target the deprecated interface continue to work, and the SDK provides default implementations of any missing functionality. After one year, the deprecated Plugin Interface may be removed in a major version release of the SDK.
```

**Review by trask — APPROVED** [review]

File: specification/upgrading.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+Likewise, application developers are encouraged to depend upon all future versions of the current major version of the OpenTelemetry API. This will ensure that version conflicts will not occur when instrumentation is upgraded to make use of new features in future minor versions.
+
+### Library Maintainers
+
+Developers who maintain OSS libraries are encouraged to use the OpenTelemetry API to add instrumentation directly into their library code. We refer to this approach as "native instrumentation." 
+
+When adding the OpenTelemetry API as a dependency to your library, it is very important to depend upon to ***all future versions*** of the current major version.
+
+Depending on a specific minor version ***will lead to version conflicts*** with other libraries that make use of API features added in later minor versions. Minor version bumps of the API will never cause a compatibility issue with existing instrumentation. It is completely safe to depend on future versions of the API.
```
Comment (tigrannajaryan) [inline specification/upgrading.md:21]: >Depending on a specific minor version ***will lead to version conflicts*** with other libraries that make use of API features added in later minor versions

I think this is language-specific. In Java the diamond dependency problem makes what you wrote true. In Go this is not needed since the compiler uses [Minimal version selection](https://go.dev/ref/mod#minimal-version-selection) logic, which given our API's backward compatibility promises will work fine without any addition measures.
It may be useful to reword this to suggest it as a possible solution, but maybe also mention Go which doesn't require this.

File: specification/upgrading.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+### SDK Implementors
+
+Maintainers of the official OpenTelemetry SDKs as well as maintainers of alternative SDKs must keep their implementations up to date and compatible with the latest version of the OpenTelemetry API.
+
+When new minor versions of the OpenTelemetry API are released, they will contain new features. This includes new interfaces, as well as new methods on current interfaces. SDK maintainers are expected to release new versions which implement these features in a timely manner.
+
+If you do not believe you will be able to continue to maintain an implementation in this manner, we strongly recommend that you do not implement an alternative SDK. It is inevitable that OpenTelemetry will release new features, and that application developers and library maintainers will make use of these features.
```
Comment (tigrannajaryan) [inline specification/upgrading.md:29]: 👍

File: specification/upgrading.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+### API changes
+
+When new functionality is added to the OpenTelemetry API, a new minor version of the API is released. These API changes are always additive and backwards compatible from the perspective of existing Instrumentation packages which import and call prior versions. Instrumentation written against all prior minor versions of the API continues to work, and may be composed together into the same application without creating a dependency conflict.
+
+API implementations are expected to always target the latest version of the API. When a new version of the API is released, a version of the SDK which supports the API is released in tandem. New versions of the API are not expected to support older versions of the SDK.
+
+### SDK changes
+
+Bugs fixes, security patches, and performance improvements are released as patch versions of the SDK. Support for new versions of the API are released as minor versions. New Plugin Interfaces and configuration options are also released as minor versions.
```
Comment (tigrannajaryan) [inline specification/upgrading.md:61]: >New Plugin Interfaces and configuration options are also released as minor versions

This part I think is a bit unclear. "minor version" of what?

File: specification/upgrading.md (hunk 1 above)
Comment (tigrannajaryan) [inline specification/upgrading.md]: >Breaking changes to Plugin Interfaces are handled through deprecation. Instead of breaking a Plugin Interface, a new interface is created and the existing interface is marked as deprecated.

This is one of the possible approaches. Another approach is to introduce a new interface that defines only the new methods, make this new interface optional for plugin implementors and probe for it at runtime. This is fairly common for e.g. Go.

I think it is OK to keep the text as is for now. I can submit a follow up PR to extend it and clarify it a bit more.

**Review by tigrannajaryan — APPROVED** [review]
LGTM, with some minor comments.

File: specification/upgrading.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+Developers who plan to install the OpenTelemetry SDK in their application are advised to depend upon all future versions of the current major version. This will ensure that applications will always be built with the latest optimizations and security patches.
+
+Likewise, application developers are encouraged to depend upon all future versions of the current major version of the OpenTelemetry API. This will ensure that version conflicts will not occur when instrumentation is upgraded to make use of new features in future minor versions.
+
+### Library Maintainers
+
+Developers who maintain OSS libraries are encouraged to use the OpenTelemetry API to add instrumentation directly into their library code. We refer to this approach as "native instrumentation." 
+
+When adding the OpenTelemetry API as a dependency to your library, it is very important to depend upon to ***all future versions*** of the current major version.
```
[the remark this answers is not in the document]
Comment (tedsuo) [reply specification/upgrading.md:19]: Yes, I understand. The problem is that exact version pinning will lead to unnecessary version conflicts between libraries. If libraries are allowed to depend on future versions, this would resolve the conflict.

If there's a better way to solve this issue, I am open to it.

File: specification/upgrading.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
+# The OpenTelemetry approach to upgrading
+
+Managing widely distributed software at scale requires careful design related to backwards compatibility, versioning, and upgrading. The OpenTelemetry approach is described below. If you are planning on using OpenTelemetry, it can be helpful to understand how we approach this problem.
+
+## Dependency Management
+It is very important that OpenTelemetry users and implementors stay up to date with the latest version of the OpenTelemetry API. OpenTelemetry follows strict rules in regards to backwards compatibility. It is always safe to upgrade to the latest minor version of an OpenTelemetry package.
+
+### Application Developers
+
+Developers who plan to install the OpenTelemetry SDK in their application are advised to depend upon all future versions of the current major version. This will ensure that applications will always be built with the latest optimizations and security patches.
```
[the remark this answers is not in the document]
Comment (tedsuo) [reply specification/upgrading.md:11]: In NPM syntax, you would use [caret ^](https://github.com/npm/node-semver#caret-ranges-123-025-004) to define this. `^1.2.3` or the equivalent `>=1.2.3 <2.0.0`.

"So, any future version of the current major version." Would it help to add the above description as a clarification?

File: specification/upgrading.md (hunk 4 above)
Comment (tedsuo) [reply specification/upgrading.md:61]: How about "New Plugin Interfaces are released as a minor version bump to the SDK."

File: specification/upgrading.md (hunk 1 above)
Comment (tedsuo) [reply specification/upgrading.md]: Thanks. What you described is what I intended to say – but with the addition that completely different interfaces are also acceptable, with a deprecation and removal policy on the old interface. When we originally wrote this, it was to keep the door open to changing the internal architecture of the SDKs. Maybe we don't care about that any more.

**Review by marcalff — COMMENTED** [review]
"New versions of the API are not expected to support older versions of the SDK."

The API does not even know what an SDK is.

Proposed rewording:

"Old versions of the SDK are not expected to support newer versions of the API."

The constraint here is on the SDK, which implements a specific API.

File: specification/upgrading.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+## The OpenTelemetry upgrade path
+
+Before we get into the design requirements, here’s how upgrading actually works. Note that all OpenTelemetry components (API, SDK, plugins, instrumentation) have separate version numbers.
+
+### API changes
+
+When new functionality is added to the OpenTelemetry API, a new minor version of the API is released. These API changes are always additive and backwards compatible from the perspective of existing Instrumentation packages which import and call prior versions. Instrumentation written against all prior minor versions of the API continues to work, and may be composed together into the same application without creating a dependency conflict.
+
+API implementations are expected to always target the latest version of the API. When a new version of the API is released, a version of the SDK which supports the API is released in tandem. New versions of the API are not expected to support older versions of the SDK.
```
Comment (marcalff) [inline specification/upgrading.md:57]: "New versions of the API are not expected to support older versions of the SDK."

The API does not even know what an SDK is.

Proposed rewording:

"Old versions of the SDK are not expected to support newer versions of the API."

The constraint here is on the SDK, which implements a specific API.

**Review by carlosalberto — APPROVED** [review]

**Comment by carlosalberto** [comment]
Overall LGTM - although I'm curious around the 1-year deprecation phase, mostly as we haven't, so far, deprecated any Plugin component.

**Review by jsuereth — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
Editorial: any chance to limit the line lengths? Commenting on a specific word/sentence is harder and github diffs are less readable with very long lines.

File: specification/upgrading.md (hunk 2 above)
Comment (tigrannajaryan) [reply specification/upgrading.md:21]: Yes, that's better.

File: specification/upgrading.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply specification/upgrading.md:11]: I understand your intent now. I am not sure how to phrase it better and concisely.

File: specification/upgrading.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+Developers who plan to install the OpenTelemetry SDK in their application are advised to depend upon all future versions of the current major version. This will ensure that applications will always be built with the latest optimizations and security patches.
+
+Likewise, application developers are encouraged to depend upon all future versions of the current major version of the OpenTelemetry API. This will ensure that version conflicts will not occur when instrumentation is upgraded to make use of new features in future minor versions.
+
+### Library Maintainers
+
+Developers who maintain OSS libraries are encouraged to use the OpenTelemetry API to add instrumentation directly into their library code. We refer to this approach as "native instrumentation."
+
+When adding the OpenTelemetry API as a dependency to your library, it is very important to depend upon to ***all future versions*** of the current major version.
```
Comment (MrAlias) [inline specification/upgrading.md:19]: Right?

```suggestion
When adding the OpenTelemetry API as a dependency to your library, it is very important to depend upon ***all future versions*** of the current major version.
```

File: specification/upgrading.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+Depending on a specific minor version ***will lead to version conflicts*** with other libraries that make use of API features added in later minor versions. Minor version bumps of the API will never cause a compatibility issue with existing instrumentation. It is completely safe to depend on future versions of the API.
+
+### SDK Implementors
+
+Maintainers of the official OpenTelemetry SDKs as well as maintainers of alternative SDKs must keep their implementations up to date and compatible with the latest version of the OpenTelemetry API.
```
Comment (MrAlias) [inline specification/upgrading.md:25]: Is this a normative statement?

```suggestion
Maintainers of the official OpenTelemetry SDKs as well as maintainers of alternative SDKs MUST keep their implementations up to date and compatible with the latest version of the OpenTelemetry API.
```

File: specification/upgrading.md [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+When new minor versions of the OpenTelemetry API are released, they will contain new features. This includes new interfaces, as well as new methods on current interfaces. SDK maintainers are expected to release new versions which implement these features in a timely manner.
+
+If you do not believe you will be able to continue to maintain an implementation in this manner, we strongly recommend that you do not implement an alternative SDK. It is inevitable that OpenTelemetry will release new features, and that application developers and library maintainers will make use of these features.
+
+## Component Overview
```
Comment (MrAlias) [inline specification/upgrading.md:31]: Shouldn't this be nearer the top given it contains term definitions that are used above?

File: specification/upgrading.md [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+Instrumentation refers to any code which calls the API. This includes the instrumentation provided by the OpenTelemetry project, third party instrumentation, plus application code and libraries which instrument themselves natively.
+
+Plugins refer to any package which implements an SDK Plugin Interface. This includes the plugins provided by the OpenTelemetry project, plus third party plugins.
+
+There is an important distinction between Plugins and Instrumentation. Plugins implement the Plugin Interfaces. Instrumentation calls the API. This difference is relevant to OpenTelemetry’s approach to upgrading.
+
+## The OpenTelemetry upgrade path
+
+Before we get into the design requirements, here’s how upgrading actually works. Note that all OpenTelemetry components (API, SDK, plugins, instrumentation) have separate version numbers.
```
Comment (MrAlias) [inline specification/upgrading.md:51]: These are capitalized proper nouns above in their definitions.

```suggestion
Before we get into the design requirements, here’s how upgrading actually works. Note that all OpenTelemetry components (API, SDK, Plugins, Instrumentation) have separate version numbers.
```

**Review by MrAlias — APPROVED** [review]

File: specification/upgrading.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (dyladan) [reply specification/upgrading.md:11]: Maybe something like "are advised to accept all minor version upgrades"
